src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Url/Base64UrlDecoder.cs

```diff
@@ -0,0 +1,777 @@
+                ];
+
+            public static ReadOnlySpan<uint> VbmiLookup0 =>
+                [
+                    0x80808080, 0x80808080, 0x80808080, 0x80808080,
+                    0x80808080, 0x80808080, 0x80808080, 0x80808080,
+                    0x80808080, 0x80808080, 0x80808080, 0x80803e80,
+                    0x37363534, 0x3b3a3938, 0x80803d3c, 0x80808080
+                ];
+
+            public static ReadOnlySpan<uint> VbmiLookup1 =>
+                [
+                    0x02010080, 0x06050403, 0x0a090807, 0x0e0d0c0b,
+                    0x1211100f, 0x16151413, 0x80191817, 0x3f808080,
+                    0x1c1b1a80, 0x201f1e1d, 0x24232221, 0x28272625,
+                    0x2c2b2a29, 0x302f2e2d, 0x80333231, 0x80808080
+                ];
+
+            public static ReadOnlySpan<sbyte> Avx2LutHigh =>
+                [
+                    0x00, 0x00, 0x2d, 0x39,
+                    0x4f, 0x5a, 0x6f, 0x7a,
+                    0x00, 0x00, 0x00, 0x00,
+                    0x00, 0x00, 0x00, 0x00,
+                    0x00, 0x00, 0x2d, 0x39,
+                    0x4f, 0x5a, 0x6f, 0x7a,
+                    0x00, 0x00, 0x00, 0x00,
+                    0x00, 0x00, 0x00, 0x00
+                ];
+
+            public static ReadOnlySpan<sbyte> Avx2LutLow =>
+                [
+                    0x01, 0x01, 0x2d, 0x30,
+                    0x41, 0x50, 0x61, 0x70,
+                    0x01, 0x01, 0x01, 0x01,
+                    0x01, 0x01, 0x01, 0x01,
+                    0x01, 0x01, 0x2d, 0x30,
+                    0x41, 0x50, 0x61, 0x70,
+                    0x01, 0x01, 0x01, 0x01,
+                    0x01, 0x01, 0x01, 0x01
+                ];
+
+            public static ReadOnlySpan<sbyte> Avx2LutShift =>
+                [
+                    0,   0,  17,   4,
+                  -65, -65, -71, -71,
+                    0,   0,   0,   0,
+                    0,   0,   0,   0,
+                    0,   0,  17,   4,
+                  -65, -65, -71, -71,
+                    0,   0,   0,   0,
+                    0,   0,   0,   0
+                ];
+
+            public static byte MaskSlashOrUnderscore => (byte)'_'; // underscore
+
+            public static ReadOnlySpan<int> Vector128LutHigh => [0x392d0000, 0x7a6f5a4f, 0x00000000, 0x00000000];
+
+            public static ReadOnlySpan<int> Vector128LutLow => [0x302d0101, 0x70615041, 0x01010101, 0x01010101];
+
+            public static ReadOnlySpan<uint> Vector128LutShift => [0x04110000, 0xb9b9bfbf, 0x00000000, 0x00000000];
+
+            public static ReadOnlySpan<uint> AdvSimdLutOne3 => [0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFF3EFF];
+
+            public static uint AdvSimdLutTwo3Uint1 => 0x1B1AFF3F;
+
+            public static int GetMaxDecodedLength(int utf8Length) => Base64Url.GetMaxDecodedLength(utf8Length);
+
+            public static bool IsInvalidLength(int bufferLength) => bufferLength % 4 == 1; // One byte cannot be decoded completely
```

Comment:
Is `bufferLength` guaranteed non-negative here?  If so, `& 3` is nicer to the CPU than `% 4`:

https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmAPgAQEwEYCwAUJgMwAEOZAwmQN5FmMXma4BsZwA9lwDZkBJFAJgA3CLygATADIBTGAHMEAC2wAKWAk4BXAGZ65AJ3lLVASjIBeAHxl1wfYZMLlKsgDIyJS1atlcAG4yAHoQsgB5GDlOAE8EGIBjCBgYLm1gGKk5RK5sqTJcsAAHXjkE3liGJlIKdk4efiERcUlZV1VNGAynY1M3XztHAz6O9wBSMgAWa38g0PComOB4pJS0jKycvLkCotLyuUqiAF8gA

```suggestion
            public static bool IsInvalidLength(int bufferLength) => (bufferLength & 3) == 1; // One byte cannot be decoded completely
```

Reply:
Yes, it is always positive